File: backend/transcoder.py
"""
视频转码 / 压缩模块

所有转码参数通过 options dict 传入，不拆分十几个独立参数。

options 结构:
{
    # ── 编码器 ──
    "video_codec":  "h264",           # h264 / h265 / vp9 / copy
    "gpu":          "CPU",            # CPU / NVIDIA / AMD / Intel
    "preset":       "medium",         # CPU 编码预设 (ultrafast ~ veryslow)

    # ── 视频参数 ──
    "resolution":   (1920, 1080),     # (width, height) 或 None = 保持
    "video_bitrate": 2000,            # kbps, None = 不设 (配合 crf)
    "framerate":    30,               # fps, None = 保持
    "crf":          None,             # CRF 质量模式 (0-51), None = 码率模式

    # ── 音频参数 ──
    "audio_codec":    "aac",          # 音频编码器, None = copy
    "audio_bitrate":  128,            # kbps
    "audio_channels": 2,              # 1 / 2 / 6, None = 保持

    # ── 其他 ──
    "extra_args": [],                 # 附加 ffmpeg 参数 (list)
    "overwrite":     True,            # 是否覆盖输出文件
}

典型调用:
    from backend.transcoder import transcode

    result = transcode(
        "input.mp4",
        "output.mp4",
        options={
            "video_codec": "h264",
            "gpu": "NVIDIA",
            "resolution": (1920, 1080),
            "video_bitrate": 3000,
            "audio_codec": "aac",
            "audio_bitrate": 128,
        },
    )
"""

from pathlib import Path

from .utils.constants import GPU_ENCODERS, HWACCEL_PARAMS
from .utils.validators import (
    validate_video_file,
    validate_resolution,
    validate_bitrate,
    validate_framerate,
    validate_output_path,
)
from .utils.exceptions import CommandBuildException, InvalidParameterException
from .executor import run_ffmpeg, TaskResult


# ─────────────────────────────────────────────
# 默认配置
# ─────────────────────────────────────────────
_DEFAULT_OPTIONS = {
    "video_codec":    "h264",
    "gpu":            "CPU",
    "preset":         "medium",
    "resolution":     None,
    "video_bitrate":  2000,
    "framerate":      None,
    "crf":            None,
    "audio_codec":    "aac",
    "audio_bitrate":  128,
    "audio_channels": 2,
    "extra_args":     [],
    "overwrite":      True,
}


# ─────────────────────────────────────────────
# 公共 helper（subtitle 模块也会调用）
# ─────────────────────────────────────────────

def _resolve_encoder(gpu: str, codec_key: str) -> str:
    """
    解析编码器名称。支持三种输入：
    1. 短键 (h264/h265/vp9) → 通过 GPU_ENCODERS 映射
    2. 完整编码器名 (libx264/h264_nvenc/mpeg4/...) → 直接使用
    3. "copy" → 流复制
    """
    if codec_key == "copy":
        return "copy"

    # 优先通过 GPU_ENCODERS 映射（短键）
    if gpu in GPU_ENCODERS and codec_key in GPU_ENCODERS[gpu]:
        return GPU_ENCODERS[gpu][codec_key]
    if codec_key in GPU_ENCODERS.get("CPU", {}):
        return GPU_ENCODERS["CPU"][codec_key]

    # 尝试作为原始编码器名直接使用
    # 常见编码器白名单（ffmpeg -encoders 输出的名称）
    _valid_encoders = {
        # H.264
        "libx264", "h264_amf", "h264_nvenc", "h264_qsv", "h264_vaapi",
        # H.265/HEVC
        "libx265", "hevc_amf", "hevc_nvenc", "hevc_qsv", "hevc_vaapi",
        # AV1
        "libaom-av1", "av1_amf", "av1_nvenc", "av1_qsv",
        # VP9
        "libvpx-vp9",
        # 其他
        "mpeg4", "prores_ks", "ffv1", "ffv1_vulkan",
        # 音频（不会被用到视频编码，但放这里安全）
        "aac", "libmp3lame", "libopus", "flac", "libfdk_aac",
    }
    if codec_key in _valid_encoders:
        return codec_key

    raise CommandBuildException(
        f"不支持的编码器: {codec_key}。支持的编码器: {sorted(_valid_encoders)}"
    )


def _build_encoding_params(options: dict) -> list[str]:
    """
    构建视频 + 音频编码参数（不含 -i / -y / 输出文件）
    返回纯参数 list，可嵌入任何 ffmpeg 命令中
    """
    opts = {**_DEFAULT_OPTIONS, **options}
    gpu = opts["gpu"]
    encoder = _resolve_encoder(gpu, opts["video_codec"])

    params = []

    # 视频编码器
    params.extend(["-c:v", encoder])

    # 分辨率
    if opts["resolution"] is not None:
        w, h = opts["resolution"]
        params.extend(["-s", f"{w}x{h}"])

    # 码率 / CRF
    if encoder != "copy":
        if opts["crf"] is not None:
            params.extend(["-crf", str(opts["crf"])])
        elif opts["video_bitrate"] is not None:
            params.extend(["-b:v", f"{opts['video_bitrate']}k"])

        # 帧率
        if opts["framerate"] is not None:
            params.extend(["-r", str(opts["framerate"])])

        # CPU 编码预设
        if gpu == "CPU" and opts["preset"]:
            params.extend(["-preset", opts["preset"]])

    # 音频
    if opts["audio_codec"] is not None:
        params.extend(["-c:a", opts["audio_codec"]])
        if opts["audio_bitrate"] is not None:
            params.extend(["-b:a", f"{opts['audio_bitrate']}k"])
        if opts["audio_channels"] is not None:
            params.extend(["-ac", str(opts["audio_channels"])])
    else:
        params.extend(["-c:a", "copy"])

    # 额外参数
    if opts.get("extra_args"):
        params.extend(opts["extra_args"])

    return params


# ─────────────────────────────────────────────
# 命令构建
# ─────────────────────────────────────────────

def build_transcode_command(
    input_path: str,
    output_path: str,
    options: dict,
) -> list[str]:
    """
    根据 options 构建 ffmpeg 命令列表
    """
    opts = {**_DEFAULT_OPTIONS, **options}

    # ── 参数校验 ──
    validate_video_file(input_path)
    validate_output_path(output_path)

    if opts["video_bitrate"] is not None:
        validate_bitrate(opts["video_bitrate"])
    if opts["framerate"] is not None:
        validate_framerate(opts["framerate"])
    if opts["resolution"] is not None:
        w, h = opts["resolution"]
        validate_resolution(w, h)

    gpu = opts["gpu"]
    encoder = _resolve_encoder(gpu, opts["video_codec"])

    # ── 组装命令 ──
    cmd = ["ffmpeg"]

    # 硬件加速解码
    if gpu in HWACCEL_PARAMS and encoder != "copy":
        cmd.extend(HWACCEL_PARAMS[gpu])

    cmd.extend(["-i", str(input_path)])

    # 视频 + 音频编码参数
    cmd.extend(_build_encoding_params(options))

    # 覆盖输出
    if opts["overwrite"]:
        cmd.append("-y")

    cmd.append(str(output_path))

    return cmd


# ─────────────────────────────────────────────
# 公共 API
# ─────────────────────────────────────────────

def transcode(
    input_path: str,
    output_path: str,
    options: dict | None = None,
    *,
    progress_callback=None,
    timeout: int = 3600,
) -> TaskResult:
    """
    视频转码 / 压缩

    Args:
        input_path: 输入视频路径
        output_path: 输出视频路径
        options: 转码参数 dict（见模块顶部结构说明）
        progress_callback: 进度回调 (stats: dict) -> None，阶段2实现
        timeout: 超时秒数（0 = 无限制）

    Returns:
        TaskResult(success, output_path, error, warnings, command, duration, log_lines)
    """
    if options is None:
        options = {}

    try:
        cmd = build_transcode_command(input_path, output_path, options)
    except CommandBuildException as e:
        return TaskResult(
            success=False,
            output_path=None,
            error=str(e),
            warnings=[],
            command=[],
            duration=0.0,
            log_lines=[],
        )
    except InvalidParameterException as e:
        return TaskResult(
            success=False,
            output_path=None,
            error=str(e),
            warnings=[],
            command=[],
            duration=0.0,
            log_lines=[],
        )
    except Exception as e:
        return TaskResult(
            success=False,
            output_path=None,
            error=f"命令构建失败: {e}",
            warnings=[],
            command=[],
            duration=0.0,
            log_lines=[],
        )

    return run_ffmpeg(cmd, progress_callback=progress_callback, timeout=timeout)

File: server_cpp/src/transcoder.cpp
#include "transcoder.h"
#include "probe.h"
#include "constants.h"
#include "installer.h"
#include <stdexcept>

namespace ffmpegpp {

// 根据编码器、编码格式和源像素格式，选择正确的输出像素格式
// H.264 编码器（libx264/h264_nvenc/h264_amf/h264_qsv）：不支持 10-bit
// H.265 编码器：libx265 支持 yuv420p10le，硬件编码器支持 p010le
static std::string resolvePixFmt(const std::string& encoder, const std::string& codec_key, const std::string& input_pix_fmt) {
    bool is10bit = (input_pix_fmt.find("10") != std::string::npos);

    // H.264 不支持 10-bit，强制降级
    if (codec_key == "h264" || encoder.find("264") != std::string::npos) {
        return "yuv420p";
    }

    // H.265 硬件编码器
    if (encoder.find("nvenc") != std::string::npos ||
        encoder.find("amf") != std::string::npos ||
        encoder.find("qsv") != std::string::npos) {
        return is10bit ? "p010le" : "nv12";
    }

    // libx265：支持 10-bit
    if (encoder == "libx265") {
        if (is10bit) {
            if (input_pix_fmt.find("422") != std::string::npos) return "yuv422p10le";
            if (input_pix_fmt.find("444") != std::string::npos) return "yuv444p10le";
            return "yuv420p10le";
        }
        return "yuv420p";
    }

    // 其他编码器：默认 8-bit
    return "yuv420p";
}

std::string resolveEncoder(const std::string& gpu, const std::string& codec_key) {
    if (codec_key == "copy") return "copy";

    // 通过 GPU_ENCODERS 映射
    if (GPU_ENCODERS.count(gpu) && GPU_ENCODERS.at(gpu).count(codec_key))
        return GPU_ENCODERS.at(gpu).at(codec_key);
    if (GPU_ENCODERS.count("CPU") && GPU_ENCODERS.at("CPU").count(codec_key))
        return GPU_ENCODERS.at("CPU").at(codec_key);

    // 直接使用原始编码器名
    static std::vector<std::string> valid = {
        "libx264", "h264_amf", "h264_nvenc", "h264_qsv",
        "libx265", "hevc_amf", "hevc_nvenc", "hevc_qsv",
        "libaom-av1", "av1_amf", "av1_nvenc", "av1_qsv",
        "libvpx-vp9", "mpeg4", "prores_ks", "ffv1",
        "aac", "libmp3lame", "libopus", "flac", "libfdk_aac",
    };
    for (auto& v : valid) {
        if (v == codec_key) return codec_key;
    }
    throw std::runtime_error("不支持的编码器: " + codec_key);
}

std::vector<std::string> buildEncodingParams(const json& options, const std::string& input_pix_fmt) {
    std::string gpu = options.value("gpu", "CPU");
    std::string video_codec = options.value("video_codec", "h264");
    bool has_vf = options.contains("vf_filters") && options["vf_filters"].is_array() && !options["vf_filters"].empty();

    std::vector<std::string> params;

    // video_codec == "none" → 纯音频模式，禁用视频流
    if (video_codec == "none") {
        params.push_back("-vn");
    } else {
        if (has_vf && video_codec == "copy") {
            video_codec = "h264";
        }
        std::string encoder = resolveEncoder(gpu, video_codec);

        params.push_back("-c:v");
        params.push_back(encoder);

        if (options.contains("pix_fmt") && !options["pix_fmt"].is_null()) {
            std::string pf = options["pix_fmt"].get<std::string>();
            if (!isInWhitelist(pf, VALID_PIX_FMTS))
                throw std::runtime_error("不支持的像素格式: " + pf);
            params.push_back("-pix_fmt");
            params.push_back(pf);
        } else if (!input_pix_fmt.empty() && encoder != "copy") {
            std::string out_fmt = resolvePixFmt(encoder, video_codec, input_pix_fmt);
            params.push_back("-pix_fmt");
            params.push_back(out_fmt);
        }

        if (options.contains("resolution") && !options["resolution"].is_null()) {
            auto res = options["resolution"];
            if (res.is_array() && res.size() == 2) {
                params.push_back("-s");
                params.push_back(std::to_string(res[0].get<int>()) + "x" + std::to_string(res[1].get<int>()));
            }
        }

        if (encoder != "copy") {
            if (options.contains("crf") && !options["crf"].is_null()) {
                params.push_back("-crf");
                params.push_back(std::to_string(options["crf"].get<int>()));
            } else if (options.contains("video_bitrate") && !options["video_bitrate"].is_null()) {
                params.push_back("-b:v");
                params.push_back(std::to_string(options["video_bitrate"].get<int>()) + "k");
            }

            if (options.contains("framerate") && !options["framerate"].is_null()) {
                params.push_back("-r");
                params.push_back(std::to_string(options["framerate"].get<double>()));
            }

            if (gpu == "CPU" && options.contains("preset")) {
                std::string pr = options["preset"].get<std::string>();
                if (!isInWhitelist(pr, VALID_PRESETS))
                    throw std::runtime_error("不支持的编码预设: " + pr);
                params.push_back("-preset");
                params.push_back(pr);
            }
        }
    }

    // 音频
    std::string audio_codec = options.value("audio_codec", "aac");
    if (!isInWhitelist(audio_codec, VALID_AUDIO_CODECS))
        throw std::runtime_error("不支持的音频编码器: " + audio_codec);
    bool has_af = options.contains("af_filters") && options["af_filters"].is_array() && !options["af_filters"].empty();
    if (has_af && (audio_codec == "copy" || audio_codec.empty())) {
        audio_codec = "aac";
    }
    if (!audio_codec.empty()) {
        params.push_back("-c:a");
        params.push_back(audio_codec);
        if (audio_codec != "copy") {
            if (options.contains("audio_bitrate") && !options["audio_bitrate"].is_null()) {
                params.push_back("-b:a");
                params.push_back(std::to_string(options["audio_bitrate"].get<int>()) + "k");
            }
            if (options.contains("sample_rate") && !options["sample_rate"].is_null()) {
                params.push_back("-ar");
                params.push_back(std::to_string(options["sample_rate"].get<int>()));
            }
        }
        if (options.contains("audio_channels") && !options["audio_channels"].is_null()) {
            params.push_back("-ac");
            params.push_back(std::to_string(options["audio_channels"].get<int>()));
        }
    } else {
        params.push_back("-c:a");
        params.push_back("copy");
    }

    return params;
}

std::vector<std::string> buildTranscodeCommand(
    const std::string& input_path,
    const std::string& output_path,
    const json& options) {

    std::string gpu = options.value("gpu", "CPU");
    std::string video_codec = options.value("video_codec", "h264");
    bool audio_only = (video_codec == "none");

    // 探测源文件像素格式（纯音频模式跳过）
    std::string input_pix_fmt;
    if (!audio_only) {
        try {
            auto probe = probeVideo(input_path);
            if (probe.success) {
                input_pix_fmt = probe.info.value("pix_fmt", "");
            }
        } catch (...) {}
    }

    std::vector<std::string> cmd = {getFFmpegPath()};

    // ── 输入/输出路径安全检查 ──
    if (!isPathSafe(input_path))
        throw std::runtime_error("输入路径包含不安全字符");
    if (!isPathSafe(output_path))
        throw std::runtime_error("输出路径包含不安全字符");

    // 片段截取：-ss 放在 -i 之前（input seeking，更快）
    if (options.contains("start_time") && !options["start_time"].is_null()) {
        cmd.push_back("-ss");
        cmd.push_back(std::to_string(options["start_time"].get<double>()));
    }

    // 硬件加速解码（纯音频模式跳过）
    if (!audio_only && HWACCEL_PARAMS.count(gpu)) {
        std::string encoder = resolveEncoder(gpu, video_codec);
        if (encoder != "copy") {
            for (auto& p : HWACCEL_PARAMS.at(gpu)) {
                cmd.push_back(p);
            }
        }
    }

    cmd.push_back("-i");
    cmd.push_back(input_path);

    // 片段截取结束时间
    if (options.contains("end_time") && !options["end_time"].is_null()) {
        cmd.push_back("-to");
        cmd.push_back(std::to_string(options["end_time"].get<double>()));
    }

    // 编码参数
    auto enc_params = buildEncodingParams(options, input_pix_fmt);
    cmd.insert(cmd.end(), enc_params.begin(), enc_params.end());

    // 视频滤镜（如变速 setpts）— 纯音频模式跳过
    if (!audio_only && options.contains("vf_filters") && options["vf_filters"].is_array() && !options["vf_filters"].empty()) {
        std::string vf;
        for (auto& f : options["vf_filters"]) {
            std::string fs = f.get<std::string>();
            if (!isFilterSafe(fs))
                throw std::runtime_error("视频滤镜包含不安全内容: " + fs);
            if (!vf.empty()) vf += ",";
            vf += fs;
        }
        cmd.push_back("-vf");
        cmd.push_back(vf);
    }

    // 音频滤镜（如变速 atempo）
    if (options.contains("af_filters") && options["af_filters"].is_array() && !options["af_filters"].empty()) {
        std::string af;
        for (auto& f : options["af_filters"]) {
            std::string fs = f.get<std::string>();
            if (!isFilterSafe(fs))
                throw std::runtime_error("音频滤镜包含不安全内容: " + fs);
            if (!af.empty()) af += ",";
            af += fs;
        }
        cmd.push_back("-af");
        cmd.push_back(af);
    }

    // 覆盖输出
    if (options.value("overwrite", true)) {
        cmd.push_back("-y");
    }

    cmd.push_back(output_path);
    return cmd;
}

} // namespace ffmpegpp
